### PR TITLE
Add TPM 2.0 emulated device option to VMIs

### PR DESCRIPTION
**Pull request by jean-edouard**
**What this PR does / why we need it**:
This PR adds an option to request a VMI to include an emulated TPM device.
This is needed for Windows 11 installation.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
TPM emulation through libvirt (prior to version 8.1.0) requires the SYS_PTRACE capability.
If we only ran as root, it would just be a matter of starting virt-launcher with that capability when TPM is enabled.
However, to be compatible with non-root, the capability has to be added to the binary, and therefore to every virt-launcher instance.
To do that, bazeldnf had to be upgraded.
Non-root compatibility also requires /var/lib/swtpm-localca to be read-writable as the qemu user.

For the functional test, the fedora containerdisk had to be updated to a version that contains the TPM2 tools.

With all that out of the way, enabling TPM is just a matter of adding the right XML block to libvirt domains.

**Release note**:

```release-note
Non-persistent vTPM now supported. Keep in mind that the state of the TPM is wiped after each shutdown. Do not enable Bitlocker!
```

### CODE REVIEW

**Comment by vasiliy-ul** [comment]
> This is just for sharing info and ideas at this point

Hi @jean-edouard . Just wanted to share info and ideas :) I remember we discussed a bit the TPM topic some time ago in [slack](https://kubernetes.slack.com/archives/C0163DT0R8X/p1639411246168000). I see in this PR you start `swtpm` manually. I guess you still have the issue with `libvirt` that it cannot start the daemon without `privileged=true` set for `virt-launcher`, right? I found that this is a bug with capabilities. I submitted a patch for `libvirt` to fix that: https://listman.redhat.com/archives/libvir-list/2022-January/msg00008.html . In short: the code in `libvirt` currently verifies that `/proc/$pid/exe` points to the correct `swtpm` binary (to ensure the daemon is still alive); however when run in a container this check requires `CAP_SYS_PTRACE` (proc(5) manpage):

> Permission to dereference or read (readlink(2)) this symbolic link is governed by a ptrace access mode PTRACE_MODE_READ_FSCREDS check; see ptrace(2).

The review of the fix is still ongoing (submitted v2 series a couple of days ago). But I hope it will be merged at some point and then it should solve the issue.

Just FYI.

**Comment by jean-edouard** [comment]
> > This is just for sharing info and ideas at this point
> 
> Hi @jean-edouard . Just wanted to share info and ideas :) I remember we discussed a bit the TPM topic some time ago in [slack](https://kubernetes.slack.com/archives/C0163DT0R8X/p1639411246168000). I see in this PR you start `swtpm` manually. I guess you still have the issue with `libvirt` that it cannot start the daemon without `privileged=true` set for `virt-launcher`, right? I found that this is a bug with capabilities. I submitted a patch for `libvirt` to fix that: https://listman.redhat.com/archives/libvir-list/2022-January/msg00008.html . In short: the code in `libvirt` currently verifies that `/proc/$pid/exe` points to the correct `swtpm` binary (to ensure the daemon is still alive); however when run in a container this check requires `CAP_SYS_PTRACE` (proc(5) manpage):
> 
> > Permission to dereference or read (readlink(2)) this symbolic link is governed by a ptrace access mode PTRACE_MODE_READ_FSCREDS check; see ptrace(2).
> 
> The review of the fix is still ongoing (submitted v2 series a couple of days ago). But I hope it will be merged at some point and then it should solve the issue.
> 
> Just FYI.

Hi @vasiliy-ul ! Thank you so much for sharing this! Congrats on finding the actual root cause in libvirt and thank you for submitting a fix!
I hope it gets accepted soon. I will test adding CAP_SYS_PTRACE to virt-launcher, which might even be an acceptable temporary solution, since we really don't want to write code to manage swtpm ourselves when libvirt already has it...
Thank you!

**Comment by vasiliy-ul** [comment]
And actually there is one more issue.. though its not very critical I think. At least it can be handled at KubeVirt side:  when libvirt starts `swtpm` daemon it uses the domain UUID to compose a unique name for the state dir. That UUID eventually appears in `/proc/$pid/cmdline` and the process monitor in virt-launcher thinks that `swtpm` is the running VM process (i.e. qemu).

I created an issue for that: https://github.com/kubevirt/kubevirt/issues/7067 but haven't got time yet to look at how to fix it.

**Comment by jean-edouard** [comment]
> And actually there is one more issue.. though its not very critical I think. At least it can be handled at KubeVirt side: when libvirt starts `swtpm` daemon it uses the domain UUID to compose a unique name for the state dir. That UUID eventually appears in `/proc/$pid/cmdline` and the process monitor in virt-launcher thinks that `swtpm` is the running VM process (i.e. qemu).
> 
> I created an issue for that: #7067 but haven't got time yet to look at how to fix it.

Noted, I will look into it!

File: cmd/virt-launcher/virt-launcher.go
Comment (vladikr) [inline cmd/virt-launcher/virt-launcher.go:431]: Perhaps it'll be better to run this in its own container and pass this socket to the launcher? 
There are a couple of immediate benefits I can think of. For example, we can manage the memory overhead separately from the known qemu overhead. It will not mess with vCPU pinning, etc... What do you think?

**Comment by vladikr** [comment]
> And actually there is one more issue.. though its not very critical I think. At least it can be handled at KubeVirt side: when libvirt starts `swtpm` daemon it uses the domain UUID to compose a unique name for the state dir. That UUID eventually appears in `/proc/$pid/cmdline` and the process monitor in virt-launcher thinks that `swtpm` is the running VM process (i.e. qemu).
> 
> I created an issue for that: #7067 but haven't got time yet to look at how to fix it.

Interesting. I think that this is another point in favor of running swtpm in a separate container.

File: cmd/virt-launcher/virt-launcher.go
Comment (jean-edouard) [reply cmd/virt-launcher/virt-launcher.go]: Interesting idea. While that would be nice to have, it would require a lot of work, and I'm not even sure we could handle things like migration without making changes to libvirt.
Thanks to Vasily's work, we can just leverage libvirt as it is today and rely on its full swtpm support, including the tricky migration.
I just updated the PR to do that, and I am currently testing it against Windows 11, which we really need to support as soon as possible...
If that works, it will mean that the small amount of code in this PR is all we'll have to maintain to support vTPM, which is a big win IMO.

**Comment by jean-edouard** [comment]
The vTPM test is expected to fail as long as that job is red (and as long as the future corresponding KubeVirt PR isn't merged):
https://prow.ci.kubevirt.io/job-history/gs/kubevirt-prow/logs/publish-kubevirtci

File: cmd/virt-launcher/virt-launcher.go
Comment (vasiliy-ul) [inline cmd/virt-launcher/virt-launcher.go:425]: Probably would be helpfull to check and log the error here.

Also when it comes to running swtpm in session mode AFAIK the config files are searched initially in the home dir. So there is an option to avoid this `Chown`  by running the script `swtpm-create-user-config-files` (it should be shipped as part of `swtpm` RPM) as `qemu` user. This script ieventually just calls `/usr/bin/swtpm_setup --create-config-files ...`. Currently this needs to be run manually but latest libvirt does that automatically https://github.com/libvirt/libvirt/commit/c66115b6e81688649da13e00093278ce55c89cb5 . Therefore with newer libvirt probably this approach will have to be revised. I am not insisting on changing smth here now. Just FYI.

File: cmd/virt-launcher/BUILD.bazel
Comment (vasiliy-ul) [inline cmd/virt-launcher/BUILD.bazel:117]: As I understand in current implementation the configs are expected to be in `os.Chown("/var/lib/swtpm-localca", 107, 107)`. Also I guess `qemu` user will not have the perms to write in `/`. Just wonder what the home dir is needed for in such case.

File: cmd/virt-launcher/virt-launcher.go
Comment (vasiliy-ul) [inline cmd/virt-launcher/virt-launcher.go:505]: nice trick :)

File: tests/vmi_tpm_test.go
Comment (vasiliy-ul) [inline tests/vmi_tpm_test.go:54]: nit: `'do''ne'` -> `done`

File: cmd/virt-launcher/virt-launcher.go
Comment (xpivarc) [inline cmd/virt-launcher/virt-launcher.go:425]: Please use an OwnershipManager as non-root can't do this.

File: cmd/virt-launcher/qemu.conf
Comment (xpivarc) [inline cmd/virt-launcher/qemu.conf:11]: Can we move this to code? The reason being we might not want to run as qemu in the future.

File: cmd/virt-launcher/virt-launcher.go
Comment (jean-edouard) [reply cmd/virt-launcher/virt-launcher.go:425]: Sounds like running that script makes more sense than chowning that directory, especially seeing the concerns related to non-root.
Hopefully we can run it as part of the virt-launcher image provisioning, so it doesn't have to happen every time we start a VMI.

File: cmd/virt-launcher/BUILD.bazel
Comment (jean-edouard) [reply cmd/virt-launcher/BUILD.bazel:117]: Probably best to create a proper home for qemu, since I assume it's needed for `swtpm-create-user-config-files` anyway.

The original reason for this change was that some auth components (not sure which), deny the existence of a system user if it doesn't have a home. For example, try `su qemu` in a compute container.
`/` is usually the no-home default, and the fact that qemu couldn't write to it was actually a good thing!

File: tests/vmi_tpm_test.go
Comment (jean-edouard) [reply tests/vmi_tpm_test.go:54]: This is actually on purpose, to avoid `BExp` matching the command instead of the result.
`SafeExpectBatch` should prevent that from happening, but I'd rather be 100% sure that if the word "done" appears anywhere in the console, it's because the `pcrextend` succeeded.

File: cmd/virt-launcher/virt-launcher.go
Comment (jean-edouard) [reply cmd/virt-launcher/virt-launcher.go:425]: Good point, but might just switch to user config, as per @vasiliy-ul suggestion, so we won't need this.

File: cmd/virt-launcher/qemu.conf
Comment (jean-edouard) [reply cmd/virt-launcher/qemu.conf:11]: I don't think we can. I'm also not sure I understand your concern. If we decide to change the user in the future, can't we just change it here?

File: cmd/virt-launcher/virt-launcher.go
Comment (jean-edouard) [reply cmd/virt-launcher/virt-launcher.go:425]: After much fighting with bazel, I was able to create a home directory for qemu...
But, even if qemu has a home, populated with `swtpm-create-user-config-files`, starting a VM still fails on trying to write to `/var/lib/swtpm-localca`... Probably need newer versions of stuff.
However, I was able to chown `/var/lib/swtpm-localca` to qemu at build time, I think it'll have to do for now, until we upgrade libvirt. Will push soon.

File: pkg/virt-controller/services/template.go
Comment (vasiliy-ul) [inline pkg/virt-controller/services/template.go:1825]: Perhaps CAP_SYS_PTRACE should go under this `if...`

File: cmd/virt-launcher/virt-launcher.go
Comment (vasiliy-ul) [reply cmd/virt-launcher/virt-launcher.go]: Hm... Interesting. Maybe some mess with env vars... Anyway, chown at build time seems like a fine approach for now.

File: pkg/virt-controller/services/template.go
Comment (jean-edouard) [reply pkg/virt-controller/services/template.go:1825]: Haha oops, gotta remove that block.
Unfortunately it can't, since the binary has the cap, the container needs it too, always...

File: cmd/virt-launcher/BUILD.bazel
Comment (vasiliy-ul) [reply cmd/virt-launcher/BUILD.bazel]: > For example, try su qemu in a compute container.

Afaik `su qemu` does not work because the user does not have a login shell defined. This can be solved  via `-s <shell>` flag:

```
84a89f842816:/ # su qemu 
This account is currently not available.
84a89f842816:/ # su qemu -s /bin/bash
qemu@84a89f842816:/> whoami
qemu
qemu@84a89f842816:/>
```

Comment (jean-edouard) [reply cmd/virt-launcher/BUILD.bazel]: Without any home dir, not even `/`?
I get:
```
$ kubectl exec -it virt-launcher-vmi-fedora-7vkc7 -c compute -- /bin/bash
bash-4.4# cat /etc/passwd 
qemu:x:107:107:user::/bin/bash
root:x:0:0:root:/root:/bin/bash
bash-4.4# su qemu
su: user qemu does not exist
bash-4.4# sed -i 's|::|:/:|' /etc/passwd
bash-4.4# su qemu
bash-4.4$ 
```

Comment (vasiliy-ul) [reply cmd/virt-launcher/BUILD.bazel]: :thinking: hm, maybe you are right. I tried in a custom container

File: tests/vmi_tpm_test.go
Comment (xpivarc) [inline tests/vmi_tpm_test.go:46]: nit: Can we use `libvmi`?

**Comment by jean-edouard** [comment]
/hold
Interesting, after the last rebase, `tpm2_pcrread sha1:16` prints "sha1:" instead of "sha1:0x0000000000000000000000000000000000000000".
That breaks the functional test. Will investigate.

**Comment by jean-edouard** [comment]
/hold cancel
So, very interesting, since this PR was first posted virt-launcher was updated and now includes swtpm 0.7.0 (instead of 0.6.z).
And it just so happens that swtpm 0.7 only enables the sha256 PCR bank (clearly stated in `swtpm_setup --help`) and leaves out the sha1 bank that was used for the test.
It makes sense, as the sha1 bank is hardly used anymore, but since it was there I initially used it for the test for shorter code lines :)
I just modified the test to use the sha256 bank instead.

File: tests/vmi_tpm_test.go
Comment (jean-edouard) [reply tests/vmi_tpm_test.go]: So, the code for this function actually uses `libvmi`, and also adds all the necessary stuff. Using `libvmi` directly I would effectively rewrite that function. Better to use it IMO.

Comment (xpivarc) [reply tests/vmi_tpm_test.go:47]: I wonder what is special about the function and how many usages it has but that's for another PR.

**Review by xpivarc — CHANGES_REQUESTED** [review]
Let`s not forget to modify SCC.

**Comment by xpivarc** [comment]
/lgtm
The only downside is that we need the capability. The code looks good :+1:

**Comment by iholder-redhat** [comment]
Thanks @jean-edouard!
/lgtm

**Comment by jean-edouard** [comment]
Force-pushed to rebase against main branch and add some verbiage to the commit message.

File: tests/libvmi/vmi.go
Comment (enp0s3) [inline tests/libvmi/vmi.go:128]: @jean-edouard I can't find the place where this function is being used, is there a chance its a leftover?

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go
Comment (enp0s3) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:1708]: @jean-edouard I think there is a mistake in the condition `(spec.Firmware.Bootloader.EFI.SecureBoot == nil || *spec.Firmware.Bootloader.EFI.SecureBoot)` as well. It will pass if no secureboot defined or if its defined and disabled, I  think we would like to check whether its not nil and enabled, or am I missing something here?

I know its not related to this PR scope, we can resolve it in a different PR if you want

File: cmd/virt-launcher/virt-launcher.go
Comment (enp0s3) [reply cmd/virt-launcher/virt-launcher.go:506]: @jean-edouard IIUC this should resolve the issue #7067, I think this change should go better in a separate commit with description about how does this change resolve it.

File: tests/vmi_tpm_test.go
Comment (enp0s3) [inline tests/vmi_tpm_test.go:70]: @jean-edouard I think its a leftover

Comment (enp0s3) [reply tests/vmi_tpm_test.go:70]: @jean-edouard Maybe you can leave a comment that explains how does the TPM functionality being checked

**Review by enp0s3 — COMMENTED** [review]
@jean-edouard Thank you! Looks good. Did a first pass, please see a few comments below.

**Comment by enp0s3** [comment]
@jean-edouard Its strange, some comments aren't being brought here, but they can be seen from the "Files Changed" tab

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go
Comment (acardace) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:1708]: @enp0s3 I don't think it's a mistake, the condition will be true if secureboot is either not defined (and we default to enabling secureBoot in this case) or defined and enabled as `spec.Firmware.Bootloader.EFI.SecureBoot` is a boolean.

Comment (enp0s3) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:1708]: @acardace Oh now I see. Thank you @acardace :)

File: tests/libvmi/vmi.go
Comment (jean-edouard) [reply tests/libvmi/vmi.go:128]: You are correct, it is not currently in use, but I thought I'd leave it there for future use.
After all, this is technically called a lib :)
Will remove on request.

File: tests/vmi_tpm_test.go
Comment (jean-edouard) [reply tests/vmi_tpm_test.go:70]: This is supposed to be the comment that explains how TPM PCR extension works, I guess it's too cryptic :)
Want me to add more verbiage?

File: cmd/virt-launcher/virt-launcher.go
Comment (jean-edouard) [reply cmd/virt-launcher/virt-launcher.go:506]: This is really a band-aid, I wouldn't say it's a proper fix for that issue...

File: tests/vmi_tpm_test.go
Comment (enp0s3) [reply tests/vmi_tpm_test.go:70]: @jean-edouard Yes I prefer to remove the cryptic comment. No need to add more comments.

File: cmd/virt-launcher/virt-launcher.go
Comment (enp0s3) [reply cmd/virt-launcher/virt-launcher.go:506]: @jean-edouard IMHO I would mark this is some way that its actually a short-term solution, and describe the actual problem its trying to solve, but its up to you, I won't insist

**Comment by enp0s3** [comment]
@jean-edouard Lets just remove the redundant comment in the e2e and push it forward. Please take a look at the downstream e2e runs as soon as this PR will be merged. I think we should skip this test there until we will adjust the virt-launcher build to have the `swtpm` installed in the container.

**Comment by enp0s3** [comment]
@jean-edouard Also we should not forget about `uuid=` prefix being a short term solution, keep tracking of [#7067](https://github.com/kubevirt/kubevirt/issues/7067)

**Comment by jean-edouard** [comment]
@enp0s3 PTAL!

File: cmd/virt-launcher/virt-launcher.go
Comment (jean-edouard) [reply cmd/virt-launcher/virt-launcher.go:511]: Added a `FIXME` comment

**Comment by jean-edouard** [comment]
@vasiliy-ul @xpivarc let me know if I haven't properly addressed one of your concerns

**Comment by enp0s3** [comment]
/approve
/hold
@jean-edouard Thank you! feel free to unhold once rebased

**Comment by enp0s3** [comment]
/unhold
@xpivarc if you don't have any other concerns, lets LGTM once rebased

File: cmd/virt-launcher/virt-launcher.go
Comment (vasiliy-ul) [inline cmd/virt-launcher/virt-launcher.go:509]: This is what libvirt currently does when looking for the swtpm process: it calls `readlink` on `/proc/$PID/exe` and that requires `CAP_SYS_PTRACE`. If we want to drop this cap in the future an alternative approach is needed. Probably a better way would be to just look in the pid-file which libvirt creates when qemu is launched (if I recall correctly that was proposed by @rmohr during one of the weekly calls).

**Review by vasiliy-ul — COMMENTED** [review]
> @vasiliy-ul @xpivarc let me know if I haven't properly addressed one of your concerns

Looks good from my perspective. Just left a small remark about the FIXME comment.

File: cmd/virt-launcher/virt-launcher.go
Comment (jean-edouard) [reply cmd/virt-launcher/virt-launcher.go]: Oh interesting, I didn't realize that's what they were doing!
Comment fixed, thanks!

File: tests/libvmi/vmi.go
Comment (xpivarc) [reply tests/libvmi/vmi.go:128]: I would probably don't introduce it as you alone are not using it.

**Comment by xpivarc** [comment]
/lgtm
/hold
@vasiliy-ul @vladikr Do you want to have final look?

File: tests/libvmi/vmi.go
Comment (jean-edouard) [reply tests/libvmi/vmi.go]: Removed

**Comment by vasiliy-ul** [comment]
> /lgtm /hold @vasiliy-ul @vladikr Do you want to have final look?

/lgtm

**Comment by vladikr** [comment]
/lgtm
Thanks @jean-edouard !

**Comment by xpivarc** [comment]
/retest
:crossed_fingers:

**Pull request merged**
